### ISSUE TITLE
Question: Cross Entropy Memory Usage

### ISSUE THREAD

**Issue by tohskai** [Dao-AILab/quack#48]
After testing quack loss on a torchtitan-esque config for Qwen3-0.6B model on H100 with torch.compile I got very stable numerics and pretty significant improvement:

<img width="1132" height="628" alt="Image" src="https://github.com/user-attachments/assets/171a9d7f-3be4-49cf-86dc-84195de063fa" />

But the memory usage got quite worse:

<img width="1132" height="618" alt="Image" src="https://github.com/user-attachments/assets/c005e901-6ca6-45dc-9a6c-429afc011ff0" />

<img width="1126" height="623" alt="Image" src="https://github.com/user-attachments/assets/5f589cf1-1a82-4308-8962-505517967855" />

One thing is that it's much bigger even after the backward finishes:

<img width="1127" height="622" alt="Image" src="https://github.com/user-attachments/assets/8d9c50bd-f2bb-4bbf-aacf-85cf01643b02" />

Are caches created by cutedsl bigger? Or the reason is something else? Any way I could try to help fix this?

**Comment by StrongerXi** [comment]
@tohskai fwiw I think `torch.compile` already does really well on `torch.nn.functional.cross_entropy`
- memory-wise it's ~0 for just fwd and ~d_input for fw+bw.
- perf-wise I don't know how it compares to quack kernels, but I think it won't amount much to the e2e perf anyways...?

**Comment by tridao** [comment]
What perf (GB/s) are you seeing w torch compile?
Qwen3-0.6B the cross entropy actually takes a significant fraction of the time.

**Comment by tohskai** [comment]
> > I think I can work on first two, but the third one seems to be hardest to implement. Do you know any examples on which I can base it around? Liger doesn't support it too afaik.
> 
> Reduction=none won't work w chunking bc chunking requires knowing what the gradient of the loss is, as it computes part of the gradient (e.g. weight grad) during the fwd. If reduction=mean or sum you know what the gradient of the loss is, it's 1.0 up to constant. What's the use case where reduction=none? If reduction is weighted sum (i.e. you know the weight of each token) then chunking will also work.

We do sum reduction, but it's done after applying multiple masks to the result. And since there are 3+ masks I can't really rely on `ignore_index`, only for padding.
